server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -484,26 +484,22 @@ public ClusterState execute(ClusterState currentState) {
                 boolean changed = false;
                 final String localNodeId = currentState.nodes().getLocalNodeId();
                 final String repoName = cloneEntry.repository();
-                final Map<String, IndexId> indexIds = getInFlightIndexIds(updatedEntries, repoName);
                 final ShardGenerations shardGenerations = repoData.shardGenerations();
                 for (int i = 0; i < updatedEntries.size(); i++) {
-                    if (cloneEntry.equals(updatedEntries.get(i))) {
+                    if (cloneEntry.snapshot().equals(updatedEntries.get(i).snapshot())) {
                         final ImmutableOpenMap.Builder<RepositoryShardId, ShardSnapshotStatus> clonesBuilder =
                                 ImmutableOpenMap.builder();
-                        // TODO: could be optimized by just dealing with repo shard id directly
-                        final Set<RepositoryShardId> busyShardsInRepo =
-                                busyShardsForRepo(repoName, snapshotsInProgress, currentState.metadata())
-                                        .stream()
-                                        .map(shardId -> new RepositoryShardId(indexIds.get(shardId.getIndexName()), shardId.getId()))
-                                        .collect(Collectors.toSet());
+                        final InFlightShardSnapshotStates inFlightShardStates =
+                            InFlightShardSnapshotStates.forRepo(repoName, snapshotsInProgress.entries());
                         for (Tuple<IndexId, Integer> count : counts) {
                             for (int shardId = 0; shardId < count.v2(); shardId++) {
                                 final RepositoryShardId repoShardId = new RepositoryShardId(count.v1(), shardId);
-                                if (busyShardsInRepo.contains(repoShardId)) {
+                                final String indexName = repoShardId.indexName();
+                                if (inFlightShardStates.isBusy(indexName, shardId)) {
                                     clonesBuilder.put(repoShardId, ShardSnapshotStatus.UNASSIGNED_QUEUED);
                                 } else {
-                                    clonesBuilder.put(repoShardId,
-                                            new ShardSnapshotStatus(localNodeId, shardGenerations.getShardGen(count.v1(), shardId)));
+                                    clonesBuilder.put(repoShardId, new ShardSnapshotStatus(localNodeId,
+                                        inFlightShardStates.bestGeneration(repoShardId.index(), shardId, shardGenerations)));
```

Comment:
Obvious bug here and in snapshot creation in hindsight ... fortunately this was not something that would corrupt the repository but it would lead to a lot of partial snapshots in practice (seen this in prod. logs as well not just this test failure).